core/node/provider.go

```diff
@@ -172,7 +172,12 @@ func SimpleProviders(reprovideStrategy string, reprovideInterval string) fx.Opti
 }
 
 func pinnedProviderStrategy(onlyRoots bool) interface{} {
-	return func(pinner pin.Pinner, dag ipld.DAGService) simple.KeyChanFunc {
-		return simple.NewPinnedProvider(onlyRoots, pinner, dag)
+	type input struct {
+		fx.In
+		Pinner      pin.Pinner
+		IPLDFetcher fetcher.Factory `name:"ipldFetcher"`
+	}
+	return func(in input) simple.KeyChanFunc {
+		return simple.NewPinnedProvider(onlyRoots, in.Pinner, in.IPLDFetcher)
```

Comment:
We are getting panics here:

```
panic: invalid key for map dagpb.PBNode: ".": no such field

goroutine 29177 [running]:
github.com/ipld/go-ipld-prime/fluent.(*mapNodeAssembler).AssembleEntry(0xc00476e290, 0x2d2b5c8, 0x1, 0x0, 0x0)
	C:/Users/adin/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.11.0/fluent/fluentBuilder.go:168 +0x1d1
github.com/ipld/go-ipld-prime/traversal/selector/builder.(*selectorSpecBuilder).Matcher.func1(0x2d74cf8, 0xc00476e290)
	C:/Users/adin/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.11.0/traversal/selector/builder/builder.go:156 +0x4f
github.com/ipld/go-ipld-prime/fluent.(*nodeAssembler).CreateMap(0xc00476e280, 0x1, 0x2cae3e0)
	C:/Users/adin/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.11.0/fluent/fluentBuilder.go:95 +0x151
github.com/ipld/go-ipld-prime/fluent.MustBuildMap.func1(0x2d856b0, 0xc00476e280)
	C:/Users/adin/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.11.0/fluent/fluentBuilder.go:31 +0x5b
github.com/ipld/go-ipld-prime/fluent.MustBuild(0x2d50660, 0x395cff0, 0xc00569fa40, 0x0, 0x0)
	C:/Users/adin/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.11.0/fluent/fluentBuilder.go:27 +0xc6
github.com/ipld/go-ipld-prime/fluent.MustBuildMap(0x2d50660, 0x395cff0, 0x1, 0x2cae3e0, 0x0, 0x0)
	C:/Users/adin/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.11.0/fluent/fluentBuilder.go:31 +0x94
github.com/ipld/go-ipld-prime/traversal/selector/builder.(*selectorSpecBuilder).Matcher(0xc00476e270, 0x0, 0x0)
	C:/Users/adin/go/pkg/mod/github.com/ipld/go-ipld-prime@v0.11.0/traversal/selector/builder/builder.go:155 +0x6b
github.com/ipfs/go-fetcher/helpers.BlockAllOfType(0x2d712d0, 0xc00946a300, 0x2d715e0, 0xc0094e3140, 0x2d63b78, 0xc00476e260, 0x2d50660, 0x395cff0, 0xc00663afe0, 0x0, ...)
	C:/Users/adin/go/pkg/mod/github.com/ipfs/go-fetcher@v1.4.0/helpers/traversal.go:46 +0xb1
github.com/ipfs/go-fetcher/helpers.BlockAll(0x2d712d0, 0xc00946a300, 0x2d715e0, 0xc0094e3140, 0x2d63b78, 0xc00476e260, 0xc00663afe0, 0x0, 0x0)
	C:/Users/adin/go/pkg/mod/github.com/ipfs/go-fetcher@v1.4.0/helpers/traversal.go:38 +0x194
github.com/ipfs/go-ipfs-provider/simple.pinSet.func1(0x2d712d0, 0xc00300aa00, 0xc008eb4250, 0x26ff2574258, 0xc00300c100, 0x2d4fe20, 0xc003000460, 0x26ff2679f00)
	C:/Users/adin/go/pkg/mod/github.com/ipfs/go-ipfs-provider@v0.6.0/simple/reprovide.go:239 +0x9a5
created by github.com/ipfs/go-ipfs-provider/simple.pinSet
	C:/Users/adin/go/pkg/mod/github.com/ipfs/go-ipfs-provider@v0.6.0/simple/reprovide.go:215 +0xbd
```

It looks like there's a bug in the fetcher helpers https://github.com/ipfs/go-fetcher/blob/ef6ef50eaf5c3968d412d87980e3ff45a7e3e6c2/helpers/traversal.go#L31-L50.

I'm not really familiar with this area of the codebase, but it looks like we're turning the selector into an IPLD node before executing it and that https://github.com/ipfs/go-fetcher/blob/ef6ef50eaf5c3968d412d87980e3ff45a7e3e6c2/helpers/traversal.go#L34 is just wrong and it needs to be a `basicnode.Prototype.Any`.

`BlockAllOfType` seems like it should just disappear as a function since it seems like the idea was that it could walk a graph if all the nodes were the same type (a bit sketchy as is), but instead the only thing the passed in `NodePrototype` does is tell selector how it should be serialized.

I'll start a new PR and reference it here.

Reply:
https://github.com/ipfs/go-fetcher/pull/24

Reply:
resolved by the above PR